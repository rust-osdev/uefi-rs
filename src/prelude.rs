//! This module is used to simplify importing the most common UEFI types.
//!
//! This includes the system table types, `Status` codes, etc.

pub use crate::{Handle, ResultExt, Status};

// Import the basic table types.
pub use crate::table::boot::BootServices;
pub use crate::table::runtime::RuntimeServices;
pub use crate::table::{Boot, SystemTable};

// Import the macro for creating the custom entry point, as well as the cstr macros.
pub use uefi_macros::{cstr16, cstr8, entry};
