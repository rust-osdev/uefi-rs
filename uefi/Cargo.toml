[package]
name = "uefi"
version = "0.21.0"
authors = ["The Rust OSDev team"]
readme = "README.md"
edition = "2021"
description = "Safe and easy-to-use wrapper for building UEFI apps."
repository = "https://github.com/rust-osdev/uefi-rs"
keywords = ["uefi", "efi"]
categories = ["embedded", "no-std", "api-bindings"]
license = "MPL-2.0"
rust-version = "1.68"

[features]
default = ["panic-on-logger-errors"]
alloc = []
global_allocator = []
logger = []
# Ignore text output errors in logger as a workaround for firmware issues that
# were observed on the VirtualBox UEFI implementation (see uefi-rs#121).
# In those cases, this feature can be excluded by removing the default features.
panic-on-logger-errors = []
# Generic gate to code that uses unstable features of Rust. You usually need a nightly toolchain.
unstable = []

[dependencies]
bitflags = "2.1.0"
log = { version = "0.4.5", default-features = false }
ptr_meta = { version = "0.2.0", default-features = false }
ucs2 = "0.3.2"
uefi-macros = "0.12.0"
uefi-raw = "0.1.0"
uguid = "2.0.0"

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]
