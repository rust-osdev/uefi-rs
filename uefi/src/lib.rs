//! Rusty wrapper for the Unified Extensible Firmware Interface.
//!
//! # Crate organisation
//!
//! The top-level module contains some of the most used types,
//! such as the result and error types, or other common data structures
//! such as GUIDs and handles.
//!
//! ## Tables and protocols
//!
//! The `table` module contains definitions of the UEFI tables,
//! which are structures containing some basic functions and references to other tables.
//! Most importantly, the boot services table also provides a way to obtain **protocol** handles.
//!
//! The `proto` module contains the standard UEFI protocols, which are normally provided
//! by the various UEFI drivers and firmware layers.
//!
//! ## Adapting to local conditions
//!
//! Unlike system tables, which are present on *all* UEFI implementations,
//! protocols *may* or *may not* be present on a certain system.
//!
//! For example, a PC with no network card might not contain a network driver,
//! therefore all the network protocols will be unavailable.

#![feature(abi_efiapi)]
#![feature(maybe_uninit_slice)]
#![feature(negative_impls)]
#![feature(ptr_metadata)]
#![cfg_attr(feature = "exts", feature(vec_into_raw_parts))]
#![cfg_attr(docsrs, feature(doc_auto_cfg))]
#![no_std]
// Enable some additional warnings and lints.
#![warn(clippy::ptr_as_ptr, missing_docs, unused)]
#![deny(clippy::all)]

// `uefi-exts` requires access to memory allocation APIs.
#[cfg(feature = "exts")]
extern crate alloc as alloc_api;

// allow referring to self as ::uefi for macros to work universally (from this crate and from others)
// see https://github.com/rust-lang/rust/issues/54647
extern crate self as uefi;

#[macro_use]
pub mod data_types;
#[cfg(feature = "exts")]
pub use self::data_types::CString16;
pub use self::data_types::{unsafe_guid, Identify};
pub use self::data_types::{CStr16, CStr8, Char16, Char8, Event, Guid, Handle};
pub use uefi_macros::guid;

mod result;
pub use self::result::{Error, Result, ResultExt, Status};

pub mod table;

pub mod proto;

pub mod prelude;

#[cfg(feature = "alloc")]
pub mod alloc;

#[cfg(feature = "logger")]
pub mod logger;
