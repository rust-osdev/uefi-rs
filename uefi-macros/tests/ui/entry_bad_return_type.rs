#![allow(unused_imports)]
#![no_main]
#![feature(abi_efiapi)]

use uefi::prelude::*;
use uefi_macros::entry;

#[entry]
fn main(_handle: Handle, _st: SystemTable<Boot>) -> bool {
    false
}
