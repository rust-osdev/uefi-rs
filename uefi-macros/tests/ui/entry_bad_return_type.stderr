error[E0308]: mismatched types
 --> tests/ui/entry_bad_return_type.rs:9:4
  |
9 | fn main(_handle: Handle, _st: SystemTable<Boot>) -> bool {
  |    ^^^^ expected struct `Status`, found `bool`
  |
  = note: expected fn pointer `extern "efiapi" fn(uefi::Handle, uefi::table::SystemTable<_>) -> Status`
                found fn item `extern "efiapi" fn(uefi::Handle, uefi::table::SystemTable<_>) -> bool {main}`
